src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs

```diff
@@ -0,0 +1,214 @@
+
+        private const string FakesConfiguratorAssembly = "Microsoft.VisualStudio.TestPlatform.Fakes";
+
+        /// <summary>
+        /// Dynamically compute the Fakes data collector settings, given a set of test assemblies
+        /// </summary>
+        /// <param name="sources">test sources</param>
+        /// <param name="runSettingsXml">runsettings</param>
+        /// <returns>updated runsettings for fakes</returns>
+        public static string GenerateFakesSettingsForRunConfiguration(string[] sources, string runSettingsXml)
+        {
+            if (sources == null)
+            {
+                throw new ArgumentNullException(nameof(sources));
+            }
+
+            if (runSettingsXml == null)
+            {
+                throw new ArgumentNullException(nameof(runSettingsXml));
+            }
+            
+            // donot generate fakes for netcore
+            if (IsNetCoreFramework(runSettingsXml))
+            {
+                return runSettingsXml;
+            }
+
+            var doc = new XmlDocument();
+            using (var xmlReader = XmlReader.Create(
+                new StringReader(runSettingsXml),
+                new XmlReaderSettings() { CloseInput = true }))
+            {
+                doc.Load(xmlReader);
+            }
+
+            return !TryAddFakesDataCollectorSettings(doc, sources) ? runSettingsXml : doc.OuterXml;
+        }
+
+        private static bool IsNetCoreFramework(string runSettingsXml)
+        {
+            var config = XmlRunSettingsUtilities.GetRunConfigurationNode(runSettingsXml);
+
+            return config.TargetFrameworkVersion.Name.IndexOf("netstandard", StringComparison.OrdinalIgnoreCase) >= 0
+                   || config.TargetFrameworkVersion.Name.IndexOf("netcoreapp", StringComparison.OrdinalIgnoreCase) >= 0;
+        }
+
+        /// <summary>
+        /// Tries to embed the Fakes data collector settings for the given run settings.
+        /// </summary>
+        /// <param name="runSettings">runsettings</param>
+        /// <param name="sources">test sources</param>
+        /// <returns>true if runSettings was modified; false otherwise.</returns>
+        private static bool TryAddFakesDataCollectorSettings(
+            XmlDocument runSettings,
+            IEnumerable<string> sources)
+        {
+            // If user provided fakes settings don't do anything
+            if (XmlRunSettingsUtilities.ContainsDataCollector(runSettings.CreateNavigator(), FakesMetadata.DataCollectorUri))
+            {
+                return false;
+            }
+
+            Func<IEnumerable<string>, string> configurator;
+
+            // fakes supported?
+            if (!TryGetFakesDataCollectorConfigurator(out configurator))
+            {
+                return false;
+            }
+
+            // if no fakes, return settings unchanged
+            var fakesConfiguration = configurator(sources);
```

Comment:
Can this configuration be done within the datacollector itself?

It will be difficult to perform such magic for external data collectors. May be our datacollector extensibility APIs require an addition to support this?